File: pwiz_tools/BiblioSpec/src/PercolatorXmlReader.cpp
//
// $Id$
//
//
// Original author: Barbara Frewen <frewen@u.washington.edu>
//
// Copyright 2012 University of Washington - Seattle, WA 98195
//
// Licensed under the Apache License, Version 2.0 (the "License"); 
// you may not use this file except in compliance with the License. 
// You may obtain a copy of the License at 
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software 
// distributed under the License is distributed on an "AS IS" BASIS, 
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
// See the License for the specific language governing permissions and 
// limitations under the License.
//

/**
 * Class definition for PercolatorXmlReader, a class for parsing the
 * xml output from percolator, versions 1.15 and later.
 */

#include <cctype>

#include "PercolatorXmlReader.h"
#include "BlibMaker.h"
#include <boost/filesystem.hpp>
#include "AminoAcidMasses.h"


namespace fs = boost::filesystem;

namespace BiblioSpec {

PercolatorXmlReader::PercolatorXmlReader(BlibBuilder& maker, 
                                         const char* file, 
                                         const ProgressIndicator* progress)
: BuildParser(maker, file, progress), 
  currentState_(START_STATE), 
  qvalueThreshold_(getScoreThreshold(SQT))
{
    this->setFileName(file); // this is done for the saxhandler
    qvalueBuffer_ = new char[256];
    qvalueBuffer_[0] = '\0';
    qvalueBufferPosition_ = qvalueBuffer_;
    AminoAcidMasses::initializeMass(masses_, 0);
}

PercolatorXmlReader::~PercolatorXmlReader()
{
    delete [] qvalueBuffer_;
}

/**
 * \brief Implementation of BuildParser virtual function.  Reads the
 * Percolator xml file and populates the vector of PSMs with the
 * information.
 */
bool PercolatorXmlReader::parseFile() {

    parse(); // the saxhandler will read the file and store the psms

    vector<std::string> extensions;
    extensions.push_back(".ms2");
    extensions.push_back(".cms2");
    extensions.push_back(".bms2");
    extensions.push_back(".pms2");

    vector<std::string> dirs;
    dirs.push_back("../sequest/");
    dirs.push_back("../");
    dirs.push_back("../../");

    BlibBuilder tmpBuilder;

    if( fileMap_.size() > 1 )
        initSpecFileProgress((int)fileMap_.size());

    // for each source file, read the .sqt for mods info, add spec to lib
    map<string, vector<PSM*> >::iterator mapAccess = fileMap_.begin();
    for(; mapAccess != fileMap_.end(); ++mapAccess){
        // filenameInput is the input source file, without the extension, may or may not be a full directory path
        string filenameInput = mapAccess->first;
        fs::path spectrumFullPath(filenameInput);
        fs::path spectrumParentPath = spectrumFullPath.parent_path();
        fs::path filename = spectrumFullPath.filename();

        if (spectrumFullPath.has_parent_path())
        {
            // if the filename contains a parent path (references other directories), extract the basename
            // and add the path to the directories to search
            dirs.push_back(spectrumParentPath.string());
        }
        // set the filename for the file containing the spectra 
        setSpecFileName(filename.string(), extensions, dirs);
        if (spectrumFullPath.has_parent_path())
        {
            // if the file had a parent path, remove that parent path from the dirs
            dirs.pop_back();
        }

        vector<PSM*>& psms = mapAccess->second;

        // create an SQTReader and get mods
        // this is the preceeding path for the spectrum file
        string fullFilename = getPath(getSpecFileName());
        // replaceExtension(filename, "sqt");
        // add on the name of the spectrum file as an sqt
        fullFilename += filename.string();
        fullFilename += ".sqt";
        ifstream file(fullFilename.c_str());

        // if the file from the path of the spectrum file can't be read
        // try opening the file in the path of the perc xml file
        if(!file.good()) {
            fullFilename = getPath(getFileName());
            fullFilename += filename.string();
            fullFilename += ".sqt";
            file.open(fullFilename.c_str());
        }

        // if the file still can't be read, try looking in the current directory
        if(!file.good()) {
            fullFilename = filename.string();
            fullFilename += ".sqt";
        }
        file.close();

        SQTreader modsReader(tmpBuilder, fullFilename.c_str(), NULL);
        try{
            modsReader.openRead(false);
        } catch(BlibException& e){ 
            const char* msg = e.what();
            if( strncmp(msg, "Couldn't open", strlen("Couldn't open")) == 0 ){
                e.addMessage(" SQT file required for reading modifications.");
                throw e;
            }// ignore warning that perc wasn't run on the sqt
        }

        applyModifications(psms, modsReader);

        psms_ = psms; // transfer to BuildParser list

        buildTables(PERCOLATOR_QVALUE);
        clearVector(psms_);
    }

    return true;
}

vector<PSM_SCORE_TYPE> PercolatorXmlReader::getScoreTypes() {
    return vector<PSM_SCORE_TYPE>(1, PERCOLATOR_QVALUE);
}

/**
 * Called when each start tag is read from the file. 
 */
void PercolatorXmlReader::startElement(const XML_Char* name,
                                       const XML_Char** attributes){

    switch(currentState_){
    case START_STATE:
        if(isElement("percolator_output", name)){
            currentState_ = ROOT_STATE;
        } 
        break;

    case ROOT_STATE:
        if(isElement("psms", name)){
            currentState_ = PSMS_STATE;
        } else if(isElement("peptides", name)){
            currentState_ = PEPTIDES_STATE;
        }
        break;

    case PSMS_STATE:
        if(isElement("psm", name)){
            parseId(attributes);
        } else if(isElement("q_value", name)){
            currentState_ = QVALUE_STATE;
            qvalueBufferPosition_ = qvalueBuffer_;//reset ptr into buffer 
        } else if(isElement("peptide_seq", name)){
            parseSequence(attributes);
        }
        break;

    case IGNORE_PSM_STATE:
    case QVALUE_STATE: //shouldn't actually get here
    case PEPTIDES_STATE:
        return;
    }

}

/**
 * Called when each closing tag is read from the file.
 */
void PercolatorXmlReader::endElement(const XML_Char* name) {
    switch(currentState_){
    case PSMS_STATE:
        if(isElement("psms", name)){
            currentState_ = ROOT_STATE;
        } else if(isElement("psm", name)){
            addCurPSM();
        }
        break;

    case IGNORE_PSM_STATE:
        if(isElement("psm", name)){
            currentState_ = PSMS_STATE;
        }
        break;
    case QVALUE_STATE:
        if(isElement("q_value", name)){
            // add the q-value to the curPSM_
            curPSM_->score = atof(qvalueBuffer_);
            if( curPSM_->score > qvalueThreshold_ ){
                currentState_ = IGNORE_PSM_STATE;
            } else {
                currentState_ = PSMS_STATE;
            }
        }
        break;
    default:
        return;
    }
}

/**
 * Given the attributes of a 'psm' tag, start a new PSM in which to
 * store data. Extract the filename, scan number, and charge from the
 * id.
 */
void PercolatorXmlReader::parseId(const XML_Char** attributes){

    const char* isdecoy = getAttrValue("p:decoy", attributes);
    if( isdecoy == NULL || strcmp("true", isdecoy)  == 0 ){
        currentState_ = IGNORE_PSM_STATE;
        return;
    }

    // if we have a current PSM, overwrite it, if not create a new one
    if( curPSM_ == NULL ){
        curPSM_ = new PSM();
    }

    // extract the filename, scan number, and charge from the id
    const char* idStr = getRequiredAttrValue("p:psm_id", attributes);
    char* buffer = new char[strlen(idStr) + 1];
    strcpy(buffer, idStr);

    // collect all tokens and interpret right to left in case filename has _
    vector<string> tokens;
    char* token = strtok(buffer, "_");
    while( token != NULL ){
        tokens.push_back(token);
        token = strtok(NULL, "_");
    }
    delete [] buffer;
    if( tokens.size() < 4 ){
        throw BlibException(false, "Error parsing psm_id '%s'.", idStr);
    }

    // tokens arranged as name, scan, charge, something
    tokens.pop_back();  // pop the last token whose meaning I've never known
    curPSM_->charge = atoi(tokens.back().c_str());  // charge
    tokens.pop_back();
    curPSM_->specKey = atoi(tokens.back().c_str()); // scan number
    tokens.pop_back();
    // reconstruct remaining tokens into a filename
    string filename = tokens.front();
    vector<string>::iterator it = tokens.begin();
    it++; // we already added the first, so skip it
    for(; it < tokens.end(); ++it){
        filename += "_";
        filename += (*it);
    }

    // hijack the specName field to store the filename
    curPSM_->specName = filename;

}

/**
 * Read the seq attribute and add it to the current PSM.
 */
void PercolatorXmlReader::parseSequence(const XML_Char** attributes){
    if(curPSM_ == NULL){
        throw BlibException(false, "Encountered a peptide sequence with no "
                            "spectrum to assign it to.");
    }

    string seq = getRequiredAttrValue("seq", attributes);
    curPSM_->unmodSeq.clear();
    // consume all bracketed mods and add them as SeqMods to the PSM
    int aaCount = 0;
    for (size_t i = 0; i < seq.length(); ++i)
    {
        char cur = seq[i];
        switch (cur)
        {
        case '[':
            {
                size_t modClose = seq.find("]", ++i);
                if (modClose == string::npos)
                {
                    throw BlibException(false, "Sequence '%s' has opening bracket "
                                        "with no closing bracket.", seq.c_str());
                }
                // get delta mass
                string modSeq = seq.substr(i, modClose - i);
                double modMass;
                try
                {
                    modMass = boost::lexical_cast<double>(modSeq);
                }
                catch (boost::bad_lexical_cast& e)
                {
                    throw BlibException(false, "Sequence '%s' has an unreadable modification. (%s)",
                                        seq.c_str(), e.what());
                }
                double deltaMass;
                // if the bracketed modification is + or - a number, assume it's a delta mass
                if (modSeq[0] == '-' || modSeq[0] == '+')
                {
                    deltaMass = modMass;
                }
                // if there is no +/- in the bracketed mass, assume it's monoisotopic residue mass + modification mass
                else
                {
                    if (i == 1)
                        throw BlibException(false, "Error assigning modification to amino acid in sequence %s", seq.c_str());
                    string residue(1, seq[i-1]);
                    double residueMass = getPeptideMass(residue, masses_);
                    deltaMass = modMass - residueMass;
                }
                // add seqmod to psm
                curPSM_->mods.push_back(SeqMod(aaCount, deltaMass));

                i = modClose;
            }
            break;
        case ']':
            {
                throw BlibException(false, "Sequence '%s' has closing bracket with no "
                                           "opening bracket.", seq.c_str());
            }
            break;
        default:
            {
                if (isalpha(cur))
                {
                    ++aaCount;
                }
                curPSM_->unmodSeq += toupper(cur);
            }
            break;
        }
    }
}

/**
 * Read all values between tags.  We only want the q-values, so only
 * read if we are in the QVALUES_STATE.
 */
void PercolatorXmlReader::characters(const XML_Char *s, int len){
    if( currentState_ != QVALUE_STATE ){
        return;
    }

    // copy len characters into current position in the qvalue buffer
    strncpy(qvalueBufferPosition_, s, len);
    qvalueBufferPosition_[len] = '\0';
    // move the current position to the end
    qvalueBufferPosition_ += len;
}

/**
 * Evaluate the current PSM and save it with the appropriate filename
 * if it passes the q-value threshold.
 */
void PercolatorXmlReader::addCurPSM(){
    if(curPSM_ == NULL)
        throw BlibException(false, "No PSM was read for this 'psm' tag.");

    if(curPSM_->score < qvalueThreshold_){ // add the psm to the map
        // retrieve and remove the filename from the psm
        string filename = curPSM_->specName;
        curPSM_->specName.clear();

        Verbosity::comment(V_DETAIL, "For file %s adding PSM: "
                           "scan %d, charge %d, sequence '%s'.",
                           filename.c_str(), curPSM_->specKey,
                           curPSM_->charge, curPSM_->unmodSeq.c_str());

        map<string, vector<PSM*> >::iterator mapAccess = 
            fileMap_.find(filename);
        if( mapAccess == fileMap_.end() ){ // not found, add the file
            vector<PSM*> tmpPsms(1, curPSM_);
            fileMap_[filename] = tmpPsms;
        } else {  // add this psm to existing file entry
            (mapAccess->second).push_back(curPSM_);
        }
        curPSM_ = NULL;
    }
}

/**
 * Use the SQTreader to translate the sequences with modification
 * symbols to a vector of mods, a seq with bracketed masses, and an
 * unmodified seq.
 */
void PercolatorXmlReader::applyModifications(vector<PSM*>& psms, 
                                             SQTreader& modsReader){
    size_t numPsms = psms.size();
    for(size_t i=0; i < numPsms; i++){
        PSM* psm = psms[i];
        // we were temporarily storing the modified seq here
        string modSeq = psm->unmodSeq;
        psm->unmodSeq.clear();
        modsReader.parseModifiedSeq(modSeq.c_str(), psm->unmodSeq, 
                                    psm->mods, false);
    }
}



} // namespace

/*
 * Local Variables:
 * mode: c
 * c-basic-offset: 4
 * End:
 */

File: pwiz-sharp/Tools/BiblioSpec/src/BiblioSpec/PercolatorXmlReader.cs
// Port of pwiz_tools/BiblioSpec/src/PercolatorXmlReader.{h,cpp}
//
// Reader for Percolator v1.15+ XML output (percolator_output / psms / psm). Each PSM carries
// a q-value (the score we keep), a peptide_seq with sequence + optional inline modifications
// (either as marker characters like '*' or as bracketed deltas like "[15.9949]"), and a
// per-file psm_id whose underscore-delimited tokens encode "<filename>_<scan>_<charge>_<idx>".
// Peaks come from the matching .ms2 / .cms2 / .bms2 / .pms2 spectrum file and the per-file
// .sqt is also consulted to translate marker characters into mod masses.

using System.Globalization;
using System.Text;
using System.Xml;

namespace Pwiz.Tools.BiblioSpec;

/// <summary>
/// Reader for Percolator XML output (<c>.perc.xml</c>) into <see cref="PSM"/>s.
/// </summary>
/// <remarks>
/// <para>Port of <c>BiblioSpec::PercolatorXmlReader</c>
/// (<c>pwiz_tools/BiblioSpec/src/PercolatorXmlReader.{h,cpp}</c>). The cpp reader is a
/// SAXHandler driven by Expat; this port walks the document with
/// <see cref="XmlReader"/>, with the same element-state state machine
/// (<see cref="ParserState"/> mirroring cpp's <c>STATE</c> enum at
/// PercolatorXmlReader.h:68).</para>
/// <para>PSMs are bucketed by source filename (parsed from <c>p:psm_id</c>) and after the
/// parse each bucket is flushed via the matching <c>.sqt</c> for mod-marker translation,
/// then handed to <see cref="BuildParser.BuildTables(PsmScoreType, string, bool, WorkflowType)"/>
/// against the bucket's spectrum file.</para>
/// </remarks>
public sealed class PercolatorXmlReader : BuildParser
{
    /// <summary>
    /// cpp parity: PercolatorXmlReader.h:68 — <c>enum STATE</c>. Drives the
    /// startElement/endElement state machine.
    /// </summary>
    private enum ParserState
    {
        Start,
        Root,            // inside <percolator_output>
        Psms,            // inside <psms>
        IgnorePsm,       // inside <psm> when decoy or q-value above threshold
        QValue,          // inside <q_value> — collect characters
        Peptides,        // inside <peptides> (we don't consume anything here)
    }

    private ParserState _state = ParserState.Start;

    // cpp parity: PercolatorXmlReader.h:74 — buffer for the characters() callback while we're
    // inside <q_value>.
    private readonly StringBuilder _qvalueBuffer = new(64);

    // cpp parity: PercolatorXmlReader.h:77 — std::map<string, vector<PSM*>>. cpp's std::map
    // iterates keys in ascending order; we use SortedDictionary so per-source flush order
    // (and the resulting library RefSpectraIDs) match cpp byte-for-byte.
    private readonly SortedDictionary<string, List<PSM>> _fileMap = new(StringComparer.Ordinal);

    // cpp parity: PercolatorXmlReader.h:78 — q-value threshold. cpp pulls from BuildInput.Sqt.
    private readonly double _qvalueThreshold;

    // cpp parity: PercolatorXmlReader.h:81 — 128-entry monoisotopic mass table for residue
    // lookups when interpreting non-delta bracketed modifications (e.g. "[15.99]" -> mass-residue).
    private readonly double[] _masses = AminoAcidMasses.BuildMassTable(monoisotopic: true);

    // Active XmlReader during parse — used by the GetAttrValue helpers. cpp's SAXHandler hides
    // this; we keep it in a field and expose the same helpers.
    private XmlReader? _reader;

    /// <summary>
    /// True if <paramref name="path"/> is a Percolator XML output file
    /// (<c>.perc.xml</c>). Used by <see cref="BlibBuilder"/>'s reader-factory dispatch.
    /// </summary>
    public static bool AcceptsExtension(string path) =>
        path.EndsWith(".perc.xml", StringComparison.OrdinalIgnoreCase);

    /// <summary>
    /// Construct a PercolatorXmlReader bound to a <see cref="BlibBuilder"/>.
    /// </summary>
    /// <remarks>cpp parity: PercolatorXmlReader.cpp:39.</remarks>
    public PercolatorXmlReader(BlibBuilder maker, string filename, ProgressIndicator? parentProgress)
        : base(maker, filename, parentProgress)
    {
        // cpp parity: PercolatorXmlReader.cpp:44 — qvalueThreshold_ comes from the SQT default,
        // because Percolator was originally an SQT-postprocessor and BlibBuilder's cutoff
        // semantics are inherited from there.
        _qvalueThreshold = GetScoreThreshold(BuildInput.Sqt);
    }

    /// <inheritdoc/>
    /// <remarks>cpp parity: PercolatorXmlReader.cpp:63.</remarks>
    public override bool ParseFile()
    {
        // cpp parity: PercolatorXmlReader.cpp:65 — saxhandler walks the XML; state machine in
        // StartElement / EndElement populates _fileMap.
        Parse();

        // cpp parity: PercolatorXmlReader.cpp:67 — spectrum-file extensions, priority order.
        var extensions = new List<string> { ".ms2", ".cms2", ".bms2", ".pms2" };

        // cpp parity: PercolatorXmlReader.cpp:73 — extra search dirs (relative to the result file).
        var dirs = new List<string> { "../sequest/", "../", "../../" };

        // cpp parity: PercolatorXmlReader.cpp:78 — a separate BlibBuilder so the modsReader
        // doesn't write into the real library.
        var tmpBuilder = new BlibBuilder();

        if (_fileMap.Count > 1)
            InitSpecFileProgress(_fileMap.Count);

        // cpp parity: PercolatorXmlReader.cpp:85 — iterate every (filename -> psms) bucket.
        foreach (var kvp in _fileMap)
        {
            // cpp parity: PercolatorXmlReader.cpp:87-105 — split the bucket name into
            // parent path + filename; if it has a parent path, prepend that to the search dirs.
            var filenameInput = kvp.Key;
            var spectrumParentPath = Path.GetDirectoryName(filenameInput) ?? string.Empty;
            var spectrumLeaf = Path.GetFileName(filenameInput);
            var hadParent = !string.IsNullOrEmpty(spectrumParentPath);

            if (hadParent)
                dirs.Add(spectrumParentPath);

            SetSpecFileName(spectrumLeaf, extensions, dirs);

            if (hadParent)
                dirs.RemoveAt(dirs.Count - 1);

            var psms = kvp.Value;

            // cpp parity: PercolatorXmlReader.cpp:109 — try the .sqt alongside the spectrum file,
            // then alongside the .perc.xml, then in the current directory.
            var fullFilename = BlibUtils.GetPath(GetSpecFileName()) + spectrumLeaf + ".sqt";
            if (!File.Exists(fullFilename))
            {
                fullFilename = BlibUtils.GetPath(GetFileName()) + spectrumLeaf + ".sqt";
                if (!File.Exists(fullFilename))
                    fullFilename = spectrumLeaf + ".sqt";
            }

            // cpp parity: PercolatorXmlReader.cpp:133 — open the SQT to learn the mod table;
            // the only failure we care about is "couldn't open" (rethrown with a hint about
            // why an SQT is needed). All other SQT parse errors are swallowed — the cpp
            // comment says "ignore warning that perc wasn't run on the sqt".
            var modsReader = new SQTreader(tmpBuilder, fullFilename, parentProgress: null);
            try
            {
                modsReader.OpenRead(warnIfNotPercolated: false);
            }
            catch (BlibException e)
            {
                if (e.Message.StartsWith("Couldn't open", StringComparison.Ordinal))
                {
                    e.AddMessage(" SQT file required for reading modifications.");
                    throw;
                }
                // ignore — perc wasn't run on the sqt
            }

            ApplyModifications(psms, modsReader);

            // cpp parity: PercolatorXmlReader.cpp:146 — transfer to BuildParser list, flush.
            Psms.Clear();
            foreach (var psm in psms)
                Psms.Add(psm);

            BuildTables(PsmScoreType.PercolatorQValue);
            Psms.Clear();
        }

        return true;
    }

    /// <inheritdoc/>
    /// <remarks>cpp parity: PercolatorXmlReader.cpp:155.</remarks>
    public override IList<PsmScoreType> GetScoreTypes() =>
        new[] { PsmScoreType.PercolatorQValue };

    // ----------------------------------------------------------------------------------
    // XmlReader driver. cpp's SAXHandler::parse() walks the document with Expat; we mirror
    // that with XmlReader, dispatching to StartElement / EndElement / Characters.
    // ----------------------------------------------------------------------------------

    private void Parse()
    {
        var settings = new XmlReaderSettings
        {
            DtdProcessing = DtdProcessing.Ignore,
            IgnoreWhitespace = false, // we need text inside <q_value>
            IgnoreComments = true,
            IgnoreProcessingInstructions = true,
            CloseInput = true,
        };

        try
        {
            using var fs = new FileStream(GetFileName(), FileMode.Open, FileAccess.Read, FileShare.Read);
            using var reader = XmlReader.Create(fs, settings);
            _reader = reader;

            while (reader.Read())
            {
                switch (reader.NodeType)
                {
                    case XmlNodeType.Element:
                    {
                        var isEmpty = reader.IsEmptyElement;
                        var name = reader.LocalName;
                        StartElement(name);
                        if (isEmpty)
                        {
                            // SAXHandler emits matching endElement for self-closing elements too.
                            EndElement(name);
                        }
                        break;
                    }
                    case XmlNodeType.EndElement:
                        EndElement(reader.LocalName);
                        break;

                    case XmlNodeType.Text:
                    case XmlNodeType.CDATA:
                    case XmlNodeType.SignificantWhitespace:
                    case XmlNodeType.Whitespace:
                        Characters(reader.Value);
                        break;
                }
            }
        }
        catch (BlibException)
        {
            throw;
        }
        catch (XmlException ex)
        {
            throw new BlibException(true,
                $"XML parse error in {GetFileName()} (line {ex.LineNumber}, position {ex.LinePosition}): {ex.Message}");
        }
        finally
        {
            _reader = null;
        }
    }

    // cpp parity: SAXHandler.h:96 — isElement (case-sensitive).
    private static bool IsElement(string expected, string actual) =>
        string.Equals(expected, actual, StringComparison.Ordinal);

    // cpp parity: SAXHandler.h:107 — getAttrValue, returns null/empty if missing.
    private string? GetAttrValue(string name) => _reader!.GetAttribute(name);

    // cpp parity: SAXHandler.h:121 — getRequiredAttrValue, throws if missing.
    private string GetRequiredAttrValue(string name)
    {
        var v = _reader!.GetAttribute(name);
        if (string.IsNullOrEmpty(v))
            throw new BlibException(true,
                $"Missing required attribute '{name}' in {GetFileName()} (element '{_reader.LocalName}').");
        return v;
    }

    // ----------------------------------------------------------------------------------
    // Element handlers — direct port of PercolatorXmlReader.cpp:162 startElement /
    // PercolatorXmlReader.cpp:202 endElement / PercolatorXmlReader.cpp:370 characters.
    // ----------------------------------------------------------------------------------

    // cpp parity: PercolatorXmlReader.cpp:162.
    private void StartElement(string name)
    {
        switch (_state)
        {
            case ParserState.Start:
                if (IsElement("percolator_output", name))
                    _state = ParserState.Root;
                break;

            case ParserState.Root:
                if (IsElement("psms", name))
                    _state = ParserState.Psms;
                else if (IsElement("peptides", name))
                    _state = ParserState.Peptides;
                break;

            case ParserState.Psms:
                if (IsElement("psm", name))
                {
                    ParseId();
                }
                else if (IsElement("q_value", name))
                {
                    _state = ParserState.QValue;
                    _qvalueBuffer.Clear();
                }
                else if (IsElement("peptide_seq", name))
                {
                    ParseSequence();
                }
                break;

            case ParserState.IgnorePsm:
            case ParserState.QValue:
            case ParserState.Peptides:
            default:
                return;
        }
    }

    // cpp parity: PercolatorXmlReader.cpp:202.
    private void EndElement(string name)
    {
        switch (_state)
        {
            case ParserState.Psms:
                if (IsElement("psms", name))
                {
                    _state = ParserState.Root;
                }
                else if (IsElement("psm", name))
                {
                    AddCurPsm();
                }
                break;

            case ParserState.IgnorePsm:
                if (IsElement("psm", name))
                {
                    _state = ParserState.Psms;
                }
                break;

            case ParserState.QValue:
                if (IsElement("q_value", name))
                {
                    // cpp parity: PercolatorXmlReader.cpp:220 — set the q-value on the curPSM_;
                    // if it's above threshold, drop into IgnorePsm for the rest of this <psm>.
                    if (!double.TryParse(_qvalueBuffer.ToString(), NumberStyles.Float,
                            CultureInfo.InvariantCulture, out var qvalue))
                    {
                        qvalue = 0; // cpp atof returns 0 on garbage
                    }
                    if (CurPsm != null)
                        CurPsm.Score = qvalue;
                    _state = CurPsm != null && CurPsm.Score > _qvalueThreshold
                        ? ParserState.IgnorePsm
                        : ParserState.Psms;
                }
                break;
        }
    }

    /// <summary>
    /// cpp parity: PercolatorXmlReader.cpp:370 — characters handler. We only want the
    /// q-value text, so we ignore everything outside <c>QValue</c> state.
    /// </summary>
    private void Characters(string text)
    {
        if (_state != ParserState.QValue || string.IsNullOrEmpty(text))
            return;
        _qvalueBuffer.Append(text);
    }

    /// <summary>
    /// cpp parity: PercolatorXmlReader.cpp:238 — psm start. Decides whether to keep this PSM
    /// based on the p:decoy attribute, and parses filename / scan / charge out of the
    /// underscore-delimited psm_id.
    /// </summary>
    private void ParseId()
    {
        // cpp parity: PercolatorXmlReader.cpp:240 — getAttrValue returns "" (not null) for
        // missing attributes, so the "p:decoy not set" path falls through. Only an explicit
        // "true" attribute trips the IgnorePsm transition.
        var isDecoy = GetAttrValue("p:decoy") ?? string.Empty;
        if (string.Equals(isDecoy, "true", StringComparison.Ordinal))
        {
            _state = ParserState.IgnorePsm;
            return;
        }

        CurPsm ??= new PSM();

        var idStr = GetRequiredAttrValue("p:psm_id");

        // cpp parity: PercolatorXmlReader.cpp:254 — split on '_' and interpret right-to-left
        // so a filename with embedded underscores still rebuilds correctly. The last token is
        // some ordinal we ignore (cpp comment: "whose meaning I've never known"), then come
        // charge, scan, and the remaining tokens form the filename.
        var tokens = idStr.Split('_');
        if (tokens.Length < 4)
            throw new BlibException(false, $"Error parsing psm_id '{idStr}'.");

        // tokens arranged: <name parts>... <scan> <charge> <ordinal>.
        // Last token is an ordinal cpp calls "whose meaning I've never known" and we ignore.
        var charge = int.TryParse(tokens[tokens.Length - 2], NumberStyles.Integer,
            CultureInfo.InvariantCulture, out var c) ? c : 0;
        var scan = int.TryParse(tokens[tokens.Length - 3], NumberStyles.Integer,
            CultureInfo.InvariantCulture, out var s) ? s : 0;

        CurPsm.Charge = charge;
        CurPsm.SpecKey = scan;

        // cpp parity: PercolatorXmlReader.cpp:275 — reconstruct the filename from the remaining
        // tokens (everything except the last three).
        var filenameTokens = tokens.Length - 3;
        var sb = new StringBuilder();
        for (var i = 0; i < filenameTokens; i++)
        {
            if (i > 0) sb.Append('_');
            sb.Append(tokens[i]);
        }
        // cpp parity: PercolatorXmlReader.cpp:284 — hijack specName to carry the filename
        // until addCurPSM moves the PSM into the file-keyed map.
        CurPsm.SpecName = sb.ToString();
    }

    /// <summary>
    /// cpp parity: PercolatorXmlReader.cpp:291 — peptide_seq handler. Strips inline bracketed
    /// mods into <see cref="PSM.Mods"/> and stores the un-bracketed sequence (still containing
    /// marker characters like '*') in <see cref="PSM.UnmodSeq"/> for later translation by
    /// <see cref="SQTreader.ParseModifiedSeq"/>.
    /// </summary>
    private void ParseSequence()
    {
        if (CurPsm is null)
        {
            throw new BlibException(false,
                "Encountered a peptide sequence with no spectrum to assign it to.");
        }

        var seq = GetRequiredAttrValue("seq");
        var sb = new StringBuilder(seq.Length);
        var aaCount = 0;

        for (var i = 0; i < seq.Length; i++)
        {
            var cur = seq[i];
            if (cur == '[')
            {
                // cpp parity: PercolatorXmlReader.cpp:308 — consume "[...]" as a bracketed mod.
                i++; // skip past '['
                var modClose = seq.IndexOf(']', i);
                if (modClose < 0)
                {
                    throw new BlibException(false,
                        $"Sequence '{seq}' has opening bracket with no closing bracket.");
                }

                var modSeq = seq.Substring(i, modClose - i);
                if (!double.TryParse(modSeq, NumberStyles.Float, CultureInfo.InvariantCulture,
                        out var modMass))
                {
                    throw new BlibException(false,
                        $"Sequence '{seq}' has an unreadable modification.");
                }

                double deltaMass;
                if (modSeq.Length > 0 && (modSeq[0] == '-' || modSeq[0] == '+'))
                {
                    // signed -> assume it's already a delta
                    deltaMass = modMass;
                }
                else
                {
                    // unsigned -> cpp PercolatorXmlReader.cpp:335 treats this as monoisotopic
                    // residue mass + modification mass and subtracts the residue mass to get the
                    // delta. cpp passes seq[i-1] which, after `++i` skipped past '[', is the
                    // '[' character itself — a latent cpp bug. cpp's getPeptideMass on "["
                    // silently returns 0 (string::find/substr accept npos), so the net effect
                    // is `deltaMass = modMass - 0`. The bracket goldens encode that. We mirror
                    // exactly: any non-A-Z residue contributes 0.
                    if (i == 1)
                    {
                        throw new BlibException(false,
                            $"Error assigning modification to amino acid in sequence {seq}");
                    }
                    var residueChar = seq[i - 1]; // cpp parity: '[' itself, not the residue
                    var residueMass = char.IsAsciiLetterUpper(residueChar)
                        ? _masses[residueChar]
                        : 0.0;
                    deltaMass = modMass - residueMass;
                }

                CurPsm.Mods.Add(new SeqMod(aaCount, deltaMass));
                i = modClose; // for-loop's ++i skips past ']'
            }
            else if (cur == ']')
            {
                throw new BlibException(false,
                    $"Sequence '{seq}' has closing bracket with no opening bracket.");
            }
            else
            {
                if (char.IsAsciiLetter(cur))
                    aaCount++;
                sb.Append(char.ToUpperInvariant(cur));
            }
        }

        CurPsm.UnmodSeq = sb.ToString();
    }

    /// <summary>
    /// cpp parity: PercolatorXmlReader.cpp:386 — psm end. If the score passed threshold,
    /// route the PSM into <see cref="_fileMap"/> keyed by its (hijacked) filename; otherwise
    /// drop it and record an empty bucket so the file is still visited by ParseFile's loop.
    /// </summary>
    private void AddCurPsm()
    {
        if (CurPsm is null)
            throw new BlibException(false, "No PSM was read for this 'psm' tag.");

        var filename = CurPsm.SpecName;

        if (CurPsm.Score < _qvalueThreshold)
        {
            // cpp parity: PercolatorXmlReader.cpp:392 — clear the hijacked field now that
            // we're filing it under the map key.
            CurPsm.SpecName = string.Empty;

            Verbosity.Comment(VerbosityLevel.Detail,
                $"For file {filename} adding PSM: scan {CurPsm.SpecKey}, charge {CurPsm.Charge}, sequence '{CurPsm.UnmodSeq}'.");

            if (!_fileMap.TryGetValue(filename, out var bucket))
            {
                bucket = new List<PSM>();
                _fileMap[filename] = bucket;
            }
            bucket.Add(CurPsm);
            CurPsm = null;
        }
        else
        {
            FilteredOutPsmCount++;
            // cpp parity: PercolatorXmlReader.cpp:413 — register the filename even though
            // we're dropping this PSM, so ParseFile still iterates the bucket.
            if (!_fileMap.ContainsKey(filename))
                _fileMap[filename] = new List<PSM>();
            CurPsm = null;
        }

        _state = ParserState.Psms;
    }

    /// <summary>
    /// cpp parity: PercolatorXmlReader.cpp:422 — translate the marker-character sequences
    /// stored in <see cref="PSM.UnmodSeq"/> into clean <see cref="PSM.UnmodSeq"/> + a list
    /// of mods, using the SQT file's DiffMod / StaticMod tables.
    /// </summary>
    private static void ApplyModifications(List<PSM> psms, SQTreader modsReader)
    {
        foreach (var psm in psms)
        {
            var modSeq = psm.UnmodSeq;
            psm.UnmodSeq = string.Empty;
            modsReader.ParseModifiedSeq(modSeq, out var cleanSeq, psm.Mods, hasFlankingAa: false);
            psm.UnmodSeq = cleanSeq;
        }
    }
}
